### PR TITLE
feat(launcher): Provide option to run puppeteer with different browsers

### PR DESCRIPTION
**Pull request by mjzffr**
This is an initial refactoring of Launcher to enable run-time browser selection in Puppeteer, starting with Firefox. On our side, we're tracking this and related work at https://bugzilla.mozilla.org/show_bug.cgi?id=1590467

/cc @whimboo @andreastt @AutomatedTester

### CODE REVIEW

**Comment by mjzffr** [comment]
@mathiasbynens Hi! This is ready for review. My CLA is in the works.

I looked at the results of the checks and it seems the failures are known to be intermittent... Let me know.

**Comment by mjzffr** [comment]
@googlebot I signed it!

**Comment by mathiasbynens** [comment]
Yay, thanks for sending this over! Just as a heads-up, I won’t be able to review this as soon as I’d like, since I’m traveling for ChromeDevSummit and BlinkOn this week. I’ll get back to y’all as soon as I can.

File: test/puppeteer.spec.js [hunk 1]
```diff
@@ -28,31 +28,40 @@ module.exports.addTests = ({testRunner, product, puppeteerPath}) => {
...
+  }, extraLaunchOptions);
+
 
   if (defaultBrowserOptions.executablePath) {
     console.warn(`${YELLOW_COLOR}WARN: running ${product} tests with ${defaultBrowserOptions.executablePath}${RESET_COLOR}`);
   } else {
-    // Make sure the `npm install` was run after the chromium roll.
-    if (!fs.existsSync(puppeteer.executablePath()))
-      throw new Error(`Browser is not downloaded. Run 'npm install' and try to re-run tests`);
+    const path = puppeteer.executablePath();
```
Comment (mathiasbynens) [inline test/puppeteer.spec.js]: Let’s use a different identifier here so as not to shadow `path` (so that it `path` remains `require('path')` throughout this entire file). How about `executablePath`?

File: lib/helper.js [hunk 2]
```diff
@@ -137,7 +137,7 @@ class Helper {
   static removeEventListeners(listeners) {
     for (const listener of listeners)
       listener.emitter.removeListener(listener.eventName, listener.handler);
-    listeners.splice(0, listeners.length);
+    listeners.length = 0;
```
Comment (mathiasbynens) [inline lib/helper.js:140]: Ooh, drive-by simplification! I dig it 👍🏻

File: test/puppeteer.spec.js [hunk 3]
```diff
@@ -28,31 +28,40 @@ module.exports.addTests = ({testRunner, product, puppeteerPath}) => {
...
+  const JUGGLER = product === 'Juggler';
 
   const puppeteer = require(puppeteerPath);
 
   const headless = (process.env.HEADLESS || 'true').trim().toLowerCase() === 'true';
   const slowMo = parseInt((process.env.SLOW_MO || '0').trim(), 10);
+  let extraLaunchOptions = {};
+  try {
+    extraLaunchOptions = JSON.parse(process.env.EXTRA_LAUNCH_OPTIONS || '{}');
+  } catch (e) {
```
Comment (mathiasbynens) [inline test/puppeteer.spec.js]: nit: `e` → `error` or `exception`

File: lib/Puppeteer.js [hunk 4]
```diff
@@ -18,6 +18,8 @@ const BrowserFetcher = require('./BrowserFetcher');
 const Errors = require('./Errors');
 const DeviceDescriptors = require('./DeviceDescriptors');
 
+const PRODUCT = process.env.PUPPETEER_PRODUCT || process.env.npm_config_puppeteer_product || process.env.npm_package_config_puppeteer_product;
```
Comment (mathiasbynens) [inline lib/Puppeteer.js]: Let’s move these `env` variables elsewhere — `puppeteer-core` must remain agnostic, and only `lib/Launcher.js` and `install.js` should consider them. `lib/Launcher.js` seems like a good fit. WDYT?

File: lib/Launcher.js [hunk 5]
```diff
@@ -294,49 +342,193 @@ class Launcher {
...
+    return firefoxArguments;
   }
 
+  /**
+   * @param {!Object=} extraPrefs
+   * @return {!Promise<string>}
+   */
+  async _createProfile(extraPrefs) {
+    const profilePath = await mkdtempAsync(path.join(os.tmpdir(), 'puppeteer_dev_firefox_profile-'));
+    const prefs_js = [];
```
Comment (mathiasbynens) [inline lib/Launcher.js]: coding style nit: `prefs_js` → `prefsJs`

same for `user_js`

File: lib/Launcher.js [hunk 6]
```diff
@@ -294,49 +342,193 @@ class Launcher {
...
+ * @param {!Puppeteer.ChildProcess} browserProcess
  * @param {number} timeout
  * @param {string} preferredRevision
  * @return {!Promise<string>}
  */
-function waitForWSEndpoint(chromeProcess, timeout, preferredRevision) {
+function waitForWSEndpoint(browserProcess, timeout, preferredRevision) {
   return new Promise((resolve, reject) => {
-    const rl = readline.createInterface({ input: chromeProcess.stderr });
+    const rl = readline.createInterface({ input: browserProcess.stderr});
```
Comment (mathiasbynens) [inline lib/Launcher.js]: nit: space before `}`

**Review by mathiasbynens — CHANGES_REQUESTED** [review]
I left some initial comments and questions. I want to do a more thorough second pass after BlinkOn, but figured this should get the conversation started.

Thanks again for your work on this!

File: lib/Puppeteer.js (hunk 4 above)
Comment (mjzffr) [reply lib/Puppeteer.js]: Yes, there may be a better place for this.

The main motivation here is to maintain some sort of privileged state about which browser we're talking to. This is both for Puppeteer API consumers and the Puppeteer implementation itself, since both internally and externally some browser-specific code may be needed from time to time.

The same holds true even in the case of `puppeteer-core`: the library doesn't download any browser binary for us, but which browser is it ultimately talking to? Or am I mistaken?

Given that, do you have any suggestions?

Comment (mjzffr) [reply lib/Puppeteer.js]: ^ @mathiasbynens 

Mentioning you in case this question got lost in a sea of notifications. :)

File: docs/api.md [hunk 7]
```diff
@@ -547,6 +549,7 @@ try {
   - `env` <[Object]> Specify environment variables that will be visible to the browser. Defaults to `process.env`.
   - `devtools` <[boolean]> Whether to auto-open a DevTools panel for each tab. If this option is `true`, the `headless` option will be set `false`.
   - `pipe` <[boolean]> Connects to the browser over a pipe instead of a WebSocket. Defaults to `false`.
+  - `extraPrefs` <[Object]> Additional [preferences](https://developer.mozilla.org/en-US/docs/Mozilla/Preferences/Preference_reference) that can be passed to Firefox (see `PUPPETEER_PRODUCT`)
```
Comment (mathiasbynens) [inline docs/api.md]: If we're gonna add a Firefox-specific option, it should probably have Firefox in its name, no?

File: lib/Puppeteer.js (hunk 4 above)
Comment (mathiasbynens) [reply lib/Puppeteer.js]: The problem is that `puppeteer-core` might be run outside of Node.js, where `process.*` is not available. In the Launcher, we wrap any usage of `process.env` in `if (!this._isPuppeteerCore) { ... }` for that reason.

File: lib/Launcher.js [hunk 8]
```diff
@@ -412,6 +604,52 @@ function getWSEndpoint(browserURL) {
...
+}
+
+/**
+ * @param {string} product
+ * @param {string} projectRoot
+ * @param {string} preferredRevision
+ * @param {boolean} isPuppeteerCore
+ * @return {ChromeLauncher|FirefoxLauncher}
+ */
+function Launcher(product, projectRoot, preferredRevision, isPuppeteerCore) {
```
Comment (mathiasbynens) [inline lib/Launcher.js]: Let's rename this `launch` now that it's no longer a class

File: lib/Puppeteer.js (hunk 4 above)
Comment (mathiasbynens) [reply lib/Puppeteer.js]: Maybe I'm misunderstanding. What problems do you see with keeping the logic local to `lib/Launcher.js`?

File: test/puppeteer.spec.js (hunk 1 above)
Comment (mjzffr) [reply test/puppeteer.spec.js]: Done.

File: docs/api.md (hunk 7 above)
Comment (mjzffr) [reply docs/api.md]: Sure. Done.

File: lib/Puppeteer.js (hunk 4 above)
Comment (mjzffr) [reply lib/Puppeteer.js]: I think I've addressed both our concerns with the latest change. :) The `PUPPETEER_PRODUCT` stuff is only used if we're not running `puppeteer-core`, and I changed Puppeteer to lazily load an appropriate Launcher based on a `product` option, if available, falling back on environment variables.

File: lib/Launcher.js (hunk 8 above)
Comment (mjzffr) [reply lib/Launcher.js]: I'm happy to, but it implies a load of type annotation changes (e.g. `Launcher.BrowserOptions`), and I'm not sure that's what you want. 

I left it as `Launcher`, but do let me know if you still want `launcher`.

**Review by mathiasbynens — APPROVED** [review]

**Review by hanselfmu — APPROVED** [review]
LGTM.

**Comment by rodoabad** [comment]
Have the documents been updated on how to use this new feature? i.e. do you still need `puppeteer-firefox`?

**Comment by chrisLovesCode** [comment]
> Have the documents been updated on how to use this new feature? i.e. do you still need `puppeteer-firefox`?

`puppeteer.launch({product: 'firefox'})`
but it's classified as experimental.
Have a look at the FAQ here: https://github.com/puppeteer/puppeteer#faq

**Comment by rodoabad** [comment]
@chrisLovesCode it doesn't say if `puppeteer` downloads firefox like how it downloads chromium or not. If we still need `puppeteer-firefox` then that's fine. Just wish the documentation was clear on that apart from explaining the new API.

**Comment by chrisLovesCode** [comment]
okay , i got your point! ..yes i agree, would be nice to get some informations, currently i  have not tried out to use firefox, but would like to do it soon, since i will write e2e tests with puppeteer

**Comment by mjzffr** [comment]
@rodoabad puppeteer doesn't download Firefox. That's tracked in https://github.com/puppeteer/puppeteer/issues/5151 

Totally agree that this should be clearer from the docs; I expect that will get better as progress is made on Firefox support.
